compiler-core/src/type_/pipe.rs

```diff
@@ -98,14 +98,13 @@ impl<'a, 'b, 'c> PipeTyper<'a, 'b, 'c> {
                     ..
                 } => {
                     let fun = self.expr_typer.infer(*fun)?;
-                    match fun.type_().fn_arity() {
-                        // Rewrite as right(left, ..args)
-                        Some(arity) if arity == arguments.len() + 1 => {
-                            self.infer_insert_pipe(fun, arguments, location)?
-                        }
-
+                    match fun.type_().fn_types() {
                         // Rewrite as right(..args)(left)
-                        _ => self.infer_apply_to_call_pipe(fun, arguments, location)?,
+                        Some((_args, return_)) if return_.fn_arity() == Some(1) => {
+                            self.infer_apply_to_call_pipe(fun, arguments, location)
+                        }
+                        // Rewrite as right(left, ..args)
+                        _ => self.infer_insert_pipe(fun, arguments, location),
```

Comment:
As we discussed earlier this is needed so that in case of a call with the wrong arity the pipe insertion `right(left, ..args)` is tried instead of `right(..args)(left)` to get better signature suggestions when someone is writing something like this: `[] |> list.map()`. This way, to the LS it looks like someone is calling `list.map([], )` and can highlight the second argument instead of being `list.map()([])` and having the first list argument highlighted

Reply:
I think this may have changed the semantics of the language and may break existing programs.

Previously if a function was missing 1 argument and returned a function that took no arguments it would rewrite to `right(left, ...args)`, but now it rewrites to `right(...args)(left)`, which is a type error.
I think this should be `if arity == arguments.len() + 1 && return_.fn_arity() == Some(1)`?

Could you add a test for this case please, to ensure the behaviour has not changed. I could be reading this wrong, so having a test to check it and to ensure it doesn't happen in future by mistake would be good.